File: Middle_LinkedList/middle_ll.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        one_step, two_step = head, head

        while two_step is not None and two_step.next is not None:
            one_step = one_step.next
            two_step = two_step.next.next

        return one_step

File: Middle_LinkedList/middle_ll.go
package middlelinkedlist

type ListNode struct {
	Val  int
	Next *ListNode
}

func middleNode(head *ListNode) *ListNode {
	oneStep, twoStep := head, head

	for twoStep != nil && twoStep.Next != nil {
		oneStep = oneStep.Next
		twoStep = twoStep.Next.Next
	}

	return oneStep
}
